src/edu/washington/escience/myria/operator/CountFilter.java

```diff
@@ -0,0 +1,213 @@
+      if (tupleIndexList == null) {
+        tupleIndexList = new IntArrayList();
+        uniqueTupleIndices.put(cntHashCode, tupleIndexList);
+      }
+      doCount.found = false;
+      doCount.meet = false;
+      doCount.sourceRow = i;
+      tupleIndexList.forEach(doCount);
+      if (!doCount.found) {
+        for (int j : keyColIndices) {
+          uniqueTuples.put(j, columns.get(j), i);
+        }
+        tupleCounts.putInt(0, 1);
+        tupleIndexList.add(nextIndex);
+        if (threshold <= 1) {
+          doCount.meet = true;
+        }
+      }
+      if (!doCount.meet) {
+        toRemove.set(i);
+      }
+    }
+    return tb.filterOut(toRemove).selectColumns(keyColIndices);
+  }
+
+  @Override
+  public Schema getSchema() {
+    Schema schema = getOp().getInputSchema();
+    if (schema == null) {
+      return null;
+    }
+    return schema.getSubSchema(keyColIndices);
+  }
+
+  @Override
+  public void init(final ImmutableMap<String, Object> execEnvVars) {
+    uniqueTupleIndices = new IntObjectHashMap<IntArrayList>();
+    uniqueTuples = new MutableTupleBuffer(getSchema());
+    tupleCounts =
+        new MutableTupleBuffer(
+            Schema.of(ImmutableList.of(Type.INT_TYPE), ImmutableList.of("count")));
+    doCount = new CountProcedure();
+  }
+
+  @Override
+  public TupleBatch update(final TupleBatch tb) {
+    TupleBatch newtb = countFilter(tb);
+    if (newtb.numTuples() > 0 || newtb.isEOI()) {
+      return newtb;
+    }
+    return null;
+  }
+
+  @Override
+  public List<TupleBatch> exportState() {
+    List<TupleBatch> tbs = uniqueTuples.getAll();
+    List<TupleBatch> ret = new ArrayList<TupleBatch>();
+    int row = 0;
+    for (TupleBatch tb : tbs) {
+      final BitSet toRemove = new BitSet(tb.numTuples());
+      for (int i = 0; i < tb.numTuples(); ++i) {
+        if (tupleCounts.getInt(0, row) < threshold) {
+          toRemove.set(i);
+        }
+        row++;
+      }
+      ret.add(tb.filterOut(toRemove));
+    }
+    return ret;
+  }
+
+  /**
+   * Traverse through the list of tuples and replace old values.
+   */
+  private transient CountProcedure doCount;
+
+  /**
+   * Traverse through the list of tuples with the same hash code.
+   */
+  private final class CountProcedure implements IntProcedure {
+
+    /** row index of the tuple. */
+    private int sourceRow;
+
+    /** input TupleBatch. */
+    private TupleBatch inputTB;
+
+    /** if found a key. */
+    private boolean found;
+
+    /** if meet the threshold for the first time. */
+    private boolean meet;
+
+    @Override
+    public void value(final int destRow) {
+      if (TupleUtils.tupleEquals(inputTB, keyColIndices, sourceRow, uniqueTuples, destRow)) {
+        found = true;
+        int oldcount = tupleCounts.getInt(0, destRow);
+        if (oldcount < threshold) {
+          tupleCounts.replace(0, destRow, new IntArrayColumn(new int[] {oldcount + 1}, 1), 0);
```

Comment:
This is a really awkward interface; I wonder how many other examples use a dummy 1-element array like this. Worth adding a new overload of `MutableTupleBuffer.replace()`?